profile-widget/divecartesianaxis.cpp

```diff
@@ -208,100 +201,142 @@ void DiveCartesianAxis::updateTicks(int animSpeed)
-		label->setZValue(1);
-		labels.push_back(label);
-		switch (position) {
-		default:
-		case Position::Bottom:
-			label->setPos(scene.sceneRect().width() + 10, rect.bottom() + labelSpaceVertical * dpr); // position it outside of the scene;
-			Animations::moveTo(labels[i], animSpeed, childPos, rect.bottom() + labelSpaceVertical * dpr);
-			break;
-		case Position::Left:
-			label->setPos(rect.left() - labelSpaceHorizontal * dpr, scene.sceneRect().height() + 10);
-			Animations::moveTo(labels[i], animSpeed, rect.left() - labelSpaceHorizontal * dpr, childPos);
-			break;
-		case Position::Right:
-			label->setPos(rect.right() + labelSpaceHorizontal * dpr, scene.sceneRect().height() + 10);
-			Animations::moveTo(labels[i], animSpeed, rect.right() + labelSpaceHorizontal * dpr, childPos);
-			break;
-		}
+		label.label = std::make_unique<DiveTextItem>(dpr, labelScale, alignFlags, this);
+		label.label->set(textForValue(value), textColor);
+		label.label->setZValue(1);
+		label.label->setPos(animSpeed <= 0 ? label.labelPosEnd : label.labelPosStart);
+		label.label->setOpacity(animSpeed <= 0 ? 1.0 : 0.0);
 	}
+	if (lineVisibility) {
+		label.lineStart = linePos(posStart);
+		label.lineEnd = linePos(pos);
+		label.line = std::make_unique<DiveLineItem>(this);
+		label.line->setPen(gridPen);
+		label.line->setZValue(0);
+		label.line->setLine(animSpeed <= 0 ? label.lineEnd : label.lineStart);
+		label.line->setOpacity(animSpeed <= 0 ? 1.0 : 0.0);
+	}
+	return label;
 }
 
-void DiveCartesianAxis::updateLines(int numTicks, double firstPosScreen, double stepScreen, int animSpeed)
+void DiveCartesianAxis::updateLabels(int numTicks, double firstPosScreen, double firstValue, double stepScreen, double stepValue,
+				     int animSpeed, double dataMinOld, double dataMaxOld)
 {
-	for (int i = 0, count = lines.size(); i < count; i++) {
-		double childPos = ((position == Position::Bottom) != inverted) ?
-					firstPosScreen + i * stepScreen :
-					firstPosScreen - i * stepScreen;
+	if (animSpeed <= 0)
+		labels.clear(); // No animation? Simply redo the labels.
+
+	std::vector<Label> newLabels;
+	newLabels.reserve(numTicks);
+	auto actOld = labels.begin();
+	double value = firstValue;
+	for (int i = 0; i < numTicks; i++, value += stepValue) {
+
+		// Check if we already got that label. If we find unused labels, mark them for deletion.
+		// Note: floating point comparisons should be fine owing to our rounding to integers above.
+		for ( ; actOld != labels.end() && actOld->value < value; ++actOld) {
+			double pos = posAtValue(actOld->value);
+			updateLabel(*actOld, 0.0, pos);
+			newLabels.push_back(std::move(*actOld));
+		}
 
-		if (position == Position::Bottom) {
-			// Fix size in case the scene changed
-			QLineF old = lines[i]->line();
-			lines[i]->setLine(old.x1(), old.y1(), old.x1(), old.y1() + rect.height());
-			Animations::moveTo(lines[i], animSpeed, childPos, rect.top());
+		double pos = ((position == Position::Bottom) != inverted) ?
```

Comment:
right... right... that's totally... obvious...
my head hurts

Reply:
I thought you would like that gnarly C-style code. ;-)

Seriously: Yes, it' really ugly, but for some reason
1) C and therefore C++ doesn't have an XOR operator.
2) Some engineer decided to make the y-coordinate for screens in opposite direction to what humanity is using since .. Descartes(??).

Reply:
because CRTs scan top down and not bottom up because of... I actually don't know

but yes, the code is fine of course. It's just not immediately intuitive.
I need to stop commenting on things that I actually agree with :)